### PR TITLE
impr: use java models for cmo generation (FPLA-1488)

### PR DESCRIPTION
**Pull request by swalker125**
### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java [hunk 1]
```diff
@@ -77,7 +77,8 @@ public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestHeader("autho
         final Map<String, Object> data = caseDetails.getData();
         final CaseData caseData = mapper.convertValue(data, CaseData.class);
 
-        Map<String, Object> cmoTemplateData = docmosisTemplateDataGenerationService.getTemplateData(caseData, true);
+        Map<String, Object> cmoTemplateData = docmosisTemplateDataGenerationService.getCaseManagementOrderData(caseData)
+            .toMap(mapper);
 
         Document document = getDocument(authorization, userId, cmoTemplateData);
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/DraftCMOController.java]: Not really sure where the best place to put this comment is but thoughts on having `AbstractDocmosisOrder` extend a base class called something like `AbstractDocmosisData` or something like that which has the default implementation of the toMap method
```java
public abstract class AbstractDocmosisData {
    public Map<String, Object> toMap(ObjectMapper mapper) {
        return mapper.convertValue(this, new TypeReference<>() {});
    }
}
```
and the the subsequent classes can call this method and perform additional logic
```java
public class AbstractDocmosisOrder extends AbstractDocmosisData {
    // fields...

    public Map<String, Object> toMap(ObjectMapper mapper) {
        Map<String, Object> map = super.toMap(mapper);
        
       // additional logic...

        return map;
    }
}
```
With this we can then have an overloaded the `DocmosisDocumentGeneratorService.generateDocmosisDocument`
to have a method that looks like
``` java
public class DocmosisDocumentGeneratorService {
    // fields and other methods...

    public DocmosisDocument generateDocmosisDocument(AbstractDocmosisData templateData, DocmosisTemplates template) {
        return generateDocmosisDocument(templateData.toMap(mapper), template);
    }
}
```
Whether we would also pass the mapper as a parameter in the method or autowire it into the class is another question as well.

I feel like this would at least from a readability point of view be an improvement as we would no longer be passing around a map that doesn't have that much meaning but instead an object which is more clear as to what it will be doing, it's use, and it's contents.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocmosisDocumentGeneratorService.java [hunk 2]
```diff
@@ -15,48 +15,46 @@
...
-
-    @Autowired
-    DocmosisDocumentGeneratorService(RestTemplate restTemplate,
-                                     DocmosisConfiguration docmosisDocumentGenerationConfiguration) {
-        this.restTemplate = restTemplate;
-        this.docmosisDocumentGenerationConfiguration = docmosisDocumentGenerationConfiguration;
+    private final DocmosisConfiguration configuration;
+    private final ObjectMapper mapper;
+
+    public DocmosisDocument generatedDocmosisDocument(AbstractDocmosisData templateData, DocmosisTemplates template) {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocmosisDocumentGeneratorService.java:30]: generate or generated? I think that generate makes more sense as that is what is actually happening.

Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/DocmosisDocumentGeneratorService.java:30]: definitely generate, typo by me

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisData.java [hunk 3]
```diff
@@ -5,8 +5,10 @@
 
 import java.util.Map;
 
-public abstract class AbstractDocmosisData {
+public class AbstractDocmosisData {
+
     public Map<String, Object> toMap(ObjectMapper mapper) {
-        return mapper.convertValue(this, new TypeReference<>() {});
+        return mapper.convertValue(this, new TypeReference<>() {
+        });
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisData.java:12]: Seems like unnecessary change

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java [hunk 4]
```diff
@@ -33,7 +34,7 @@
 
     @JsonIgnore
     public boolean isDraft() {
-        return action == null || !action.getType().equals(SEND_TO_ALL_PARTIES);
+        return ofNullable(action).map(x -> !SEND_TO_ALL_PARTIES.equals(action.getType())).orElse(true);
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java:37]: Not sure then conversion to the optional improves anything although I would agree that the flipping of the equals is better

Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java:37]: If you want to be insistent on the optional then it would be better to use x in the map function rather than action

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersControllerMidEventTest.java [hunk 5]
```diff
@@ -55,24 +56,24 @@
...
+            .willReturn(docmosisDocument);
 
-    @Test
-    void midEventShouldGenerateDraftStandardDirectionDocument() {
         given(uploadDocumentService.uploadPDF(PDF, DRAFT_ORDER_FILE_NAME))
             .willReturn(document);
+    }
 
+    @Test
+    void midEventShouldGenerateDraftStandardDirectionDocument() {
```
Comment (lewisbirks) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersControllerMidEventTest.java:76]: Don't think that the test needs to start with `midEventShould`, just `should` should be enough

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisData.java (hunk 3 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisData.java:12]: ahh damn, reformat code does this for me by default, shall revert. Good spot

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java (hunk 4 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseManagementOrder.java:37]: fine to revert and then swap the equals around

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersControllerMidEventTest.java (hunk 5 above)
Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/DraftOrdersControllerMidEventTest.java:76]: I totally agree, however this was done in a refactoring of draft orders controller and not in this PR. I am not going address within this.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisCaseManagementOrder.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    private final Schedule schedule;
+    private final boolean scheduleProvided;
+
+    private final List<DocmosisRecital> recitals;
+
+    //not needed can manage in doc
+    private final boolean recitalsProvided;
+
+    //not needed can manage in doc
+    private final int numberOfChildren;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisCaseManagementOrder.java:21]: is this going to be done in this PR or in a future date?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java [hunk 7]
```diff
@@ -61,8 +61,7 @@
     @Autowired
     private HearingVenueLookUpService hearingVenueLookUpService;
 
-    private CommonCaseDataExtractionService commonCaseDataExtractionService = new CommonCaseDataExtractionService(
-        hearingVenueLookUpService);
+    private CommonCaseDataExtractionService commonCaseDataExtractionService = new CommonCaseDataExtractionService();
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java:64]: Could just be autowired in

Comment (lewisbirks) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java:64]: Although I guess just creating a new one doesn't actually hurt and requires less from spring so may be slightly better in performance

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderControllerAboutToSubmitTest.java [hunk 8]
```diff
@@ -68,6 +72,7 @@
     private static final UUID ID = randomUUID();
 
     private final DocumentReference cmoDocument = buildFromDocument(document());
+    private CaseDetails populatedCaseDetails;
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderControllerAboutToSubmitTest.java:75]: this could be private final CaseDetails populatedCaseDetails  = populatedCaseDetails();

and removed from setup

Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderControllerAboutToSubmitTest.java:75]: I update populatedCaseDetails in each test, so needs to be reset in the beforeEach

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisCaseManagementOrder.java (hunk 6 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisCaseManagementOrder.java:21]: doing now!

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java (hunk 7 above)
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java:64]: will leave as is

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderControllerAboutToSubmitTest.java (hunk 8 above)
Comment (swalker125) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderControllerAboutToSubmitTest.java:75]: I update populatedCaseDetails directly so I need to reset before each test in the before each

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+import static java.util.stream.Collectors.groupingBy;
+import static org.apache.commons.lang3.ObjectUtils.isNotEmpty;
+
+@Getter
+@AllArgsConstructor
+public class AbstractDocmosisOrder extends AbstractDocmosisData {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:15]: Are we sure all of those are true for any order? I.e. we don't have complianceDeadline for most of the order do we? It seems only for SDO / CMO, might be confusing if we wanted to reuse this class for other orders. Would you consider renaming?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    private final DocmosisJudgeAndLegalAdvisor judgeAndLegalAdvisor;
+    private final String courtName;
+    private final String familyManCaseNumber;
+    private final String dateOfIssue;
+    private final String complianceDeadline;
+    private final List<DocmosisRespondent> respondents;
+    private final List<DocmosisChild> children;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:22]: Probably not ideal with duplicating the Docmosis from the package name, but I can imagine it makes development easier due to naming clash, so it's fine.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    private final boolean respondentsProvided;
+    private final String applicantName;
+    private final DocmosisHearingBooking hearingBooking;
+    private final List<DocmosisDirection> directions;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:26]: Again not every order has direction, this is abstract, but not really modelling the reality :(

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisStandardDirectionOrder.java [hunk 12]
```diff
@@ -1,42 +1,28 @@
...
-
-        if (isNotEmpty(this.directions)) {
-            map.putAll(this.directions.stream().collect(groupingBy(direction -> direction.assignee.getValue())));
-        }
-
-        map.remove("directions");
-
-        return map;
+@Getter
+public class DocmosisStandardDirectionOrder extends AbstractDocmosisOrder {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisStandardDirectionOrder.java:9]: Maybe the order doesn't need to be abstract then? This class doesn't provide much value.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java [hunk 13]
```diff
@@ -1,7 +1,6 @@
 package uk.gov.hmcts.reform.fpl.utils;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java:1]: Shouldn't this leave with tests?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderGenerationServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,287 @@
+package uk.gov.hmcts.reform.fpl.service;
```
Comment (akomisarek) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderGenerationServiceTest.java:1]: Let's ask Raghu/Dough to review it :D

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java (hunk 9 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:15]: No these are not all common for orders, I've already had this chat with Lewis and I going to keep working on as I work my way through other orders. This Order object will probably turn into something a little more specific to group is as CaseManagementOrders maybe? And AbstractDocmosisOrder will keep whatever is common.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java (hunk 10 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:22]: Yes, so I don't really want Docmosis in the name at all, I just think it's confusing to be using two different Child types in the same lines of code. Any other ideas for this? Maybe being the same name is fine?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java (hunk 11 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:26]: related to top comment

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java (hunk 13 above)
Comment (lewisbirks) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java:1]: I think so but I think it also used within both integration and unit tests so one breaks if it is placed in the package of the other

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisStandardDirectionOrder.java (hunk 12 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/DocmosisStandardDirectionOrder.java:9]: again, I believe related to my top comment.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java (hunk 13 above)
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java:1]: YES. I despise this class. However, it is used in integration AND unit tests so isn't as simple as just moving. I am not addressing within this PR

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderGenerationServiceTest.java (hunk 14 above)
Comment (swalker125) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/CaseManagementOrderGenerationServiceTest.java:1]: yep, I am going to move it to QA for a sanity check

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java (hunk 10 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/docmosis/AbstractDocmosisOrder.java:22]: It's fine, but Java will force you to use fully typed name for one of them so you will have:
uk.gov.hmcts.reform.fpl.model.docmosis.Child which is a bit annoying as well. So you can import only one with clashing name and need to fully reference the other. 

I personally prefer fully referencing the less used one as it is anyway only one or two lines where the mapping happens, so maybe something you could consider.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java (hunk 13 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/utils/CaseDataGeneratorHelper.java:1]: Leave means live! :D but yes happy to not address it here.
